libsbom/spdx.py

```diff
@@ -0,0 +1,213 @@
+                      name=f"alma-sbom {__version__}")
+
+    return Annotation(spdx_id=spdxid,
+                      annotation_type=AnnotationType.OTHER,
+                      annotator=this_tool,
+                      annotation_date=datetime.datetime.now(),
+                      annotation_comment=content)
+
+
+def make_checksum(algo: str, value: str) -> Checksum:
+    algo_map = {
+        "SHA-256": ChecksumAlgorithm.SHA256
+    }
+
+    if not algo in algo_map:
+        raise ValueError(f"Algorithm {algo} is not supported")
+
+    return Checksum(algo_map[algo], value)
+
+
+def component_get_buildtime(component: dict) -> datetime.datetime:
+    buildtime = component_get_property(component, "almalinux:package:timestamp")
+    return datetime.datetime.fromisoformat(buildtime)
+
+
+class SBOM:
+    def __init__(self, data, sbom_object_type, output_format, output_file):
+        self._input_data = data
+        self._sbom_object_type = sbom_object_type
+        self._output_format = output_format
+        self._output_file = output_file
+
+        self._org = Actor(ActorType.ORGANIZATION,
+                          ALMAOS_VENDOR,
+                          ALMAOS_EMAIL)
+        self._creators = [self._org]
+
+        for tool in TOOLS:
+            self._creators += [Actor(actor_type=ActorType.TOOL,
+                                     name=f"{tool['name']} {tool['version']}")]
+
+        self._document = None
+        self._next_id = 0
+        self._prepare_document()
+
+
+    def get_next_package_id(self) -> str:
+        """Return an identifier that can be assigned to a package in this document.
+
+        Further reading:
+        https://spdx.github.io/spdx-spec/v2-draft/package-information/#72-package-spdx-identifier-field
+        """
+        cur_id = self._next_id
+        self._next_id += 1
+        return f"SPDXRef-{cur_id}"
+
+
+    def _prepare_document(self):
+        if "metadata" in self._input_data:
+            doc_name = self._input_data["metadata"]["name"]
+        else:
+            pkgname = self._input_data['component']['name']
+            pkgvers = self._input_data['component']['version']
+            doc_name = f"{pkgname}-{pkgvers}"
+
+        doc_info = CreationInfo(spdx_version="SPDX-2.3",
+                                spdx_id=f"SPDXRef-{uuid.uuid4()}",
+                                name=doc_name,
+                                data_license=ALMAOS_SBOMLICENSE,
+                                document_namespace=ALMAOS_NAMESPACE,
+                                creators=self._creators,
+                                created=datetime.datetime.now())
+
+        self._document = Document(doc_info)
+
+
+    def add_package(self, component):
+        pkgid = self.get_next_package_id()
+
+        pkg = Package(spdx_id=pkgid,
+                      name=component["name"],
+                      download_location=component["purl"]) # FIXME: download_location must point to the RPM on a mirror
```

Comment:
I guess you can use a direct link to ALBS (use open API and get link to a package)
https://build.almalinux.org/api/v1/builds/<build_id>/

Reply:
I just had a look at the API. Can I use something like the following for the `download_location` field? (I'm asking because I'm getting a 403 on the resource and I don't know how to authenticate with pulp).

```
https://build.almalinux.org/pulp/api/v3/content/rpm/packages/b927446c-09ed-4f7f-9b4b-22e22ce40ab4/
```

Reply:
I think you can concatenate an URL to a package using info from the endpoint above.
It looks like 

```
https://build.almalinux.org/pulp/content/builds/<platform_name>-<"src" or arch>-<build_id>-br/Packages/<first_letter_of_package>/<package_name>
```
As I see all fields can be extracted from the endpoint's response, like "https://build.almalinux.org/api/v1/builds/1372/"

Reply:
I like your idea @soksanichenko, however, we don't provide the platform IIRC, right? That would probably involve poking ALBS API to fetch such info, for which we would need to provide an access token. Am I right?
That said, bear in mind that, from the docs, the field `downloadLocation` is not mandatory and it can be set to null.
Also, note that the docs say they following about the downloadLocation: `... identifies the download Uniform Resource Locator (URL), or a specific location within a version control system (VCS) for the package at the time that the SPDX document was created.`. I'd like to particularly highlight **at the time that the SPDX document was created**.  Since we do not automatically create SBOMs right now, I'd say we're okay setting it to `NONE`. What do you @soksanichenko and @m10k think?

Reply:
It's not needed to provide a token, because the build feed is public and other related stuff too.
I suggested that because want to save the same logic as for CycloneDX format. There is we provide build URL.
Anyway, it's up to you and @m10k =)

Reply:
> It's not needed to provide a token, because the build feed is public and other related stuff too.

That's right, `builds` endpoint returns the platform name and we don't need anything else.

> I suggested that because want to save the same logic as for CycloneDX format. There is we provide build URL. Anyway, it's up to you and @m10k =)

Well, technically speaking, we don't deliver a specific URL (or download location) but a `purl`, i.e.: `pkg:rpm/almalinux/bash@5.1.8-6.el9_1?arch=x86_64&upstream=bash-5.1.8-6.el9_1.src.rpm`
That's my reasoning for setting `downloadLocation` to null.

Reply:
Ok. let `downloadLocation` be empty.